Modify enum default generation

joaommartins opened the pull request:
Fix for #136, refactoring `value_for_external_enum` and separating out the enum variant and value extraction into a function usable in other contexts.

The context that requires this is `impl Default for <Enum>` , where we can make use of this function to extract the variant name for simple, and only simple, enums variants.

ahl commented:
I had been looking at a related / overlapping problem for which I was considering changing `output_value` to take a parameter denoting a "path prefix". I think that *all* the calls to `output_value` from `type_entry.rs` suffer from this problem. The original use was for serde default functions which are generated in their own `mod defaults`; the additional use of `output_value` in other places failed to consider this, and only works on built-in types such as `u32`.

Could I ask you to consider that alternative approach? If it turns out to be worse, I'm certainly open to the approach you've taken here. Thanks!

joaommartins commented:
@ahl  I've updated the PR to reflect your suggested changes, `output_value` now has a bool argument for generating super-scoped implementations or same-scoped.

From what I could gather, the super scope is only used in the serde default generation, but I've added tests to struct and enum `output_value` calls with both scopes for coverage purposes to capture possible regressions later on.

ahl commented on typify-impl/src/value.rs:
```diff
@@ -25,6 +26,7 @@ impl TypeEntry {
         &self,
         type_space: &TypeSpace,
         value: &serde_json::Value,
+        super_scoped: bool,
```
What would you think about making this into a `scope: TokenStream`?

joaommartins replied:
Reduces code duplication, it's a fair point, I'll get that done.

ahl commented on typify-impl/src/enums.rs:
```diff
@@ -1453,6 +1453,71 @@ mod tests {
     }
 
+    #[test]
+    fn test_enum_default_generation_simple_variant() {
```
I recently added a test harness that converts json schema files into .rs files and compiles them. Could you just add this as a file in the typify/schemas/ directory? I think it will be equivalent to what you have **and** will compile the resulting output.

ahl replied:
Can you please also add the test case from #125 ? I think it should be fixed by this now as well.

ahl replied:
Nevermind: I fixed that case differently

ahl replied:
but please DO put THIS new test into a file in typify/tests/schemas

joaommartins commented:
Bool check removed and replaced with direct TokenStream use.

ahl commented on typify-impl/src/value.rs:
```diff
@@ -34,53 +36,53 @@ impl TypeEntry {
             TypeEntryDetails::Struct(TypeEntryStruct {
                 name, properties, ..
             }) => {
                 let props = value_for_struct_props(properties, value, type_space)?;
```
why does this not require scope?

ahl commented on typify-impl/src/value.rs:
```diff
@@ -393,13 +404,13 @@ mod tests {
 
```
I think using scope = quote!{ xxx }; would be a better test of all these cases

ahl commented on typify-impl/src/value.rs:
```diff
@@ -773,7 +1025,11 @@ mod tests {
+                .output_value(
+                    &type_space,
+                    &json!(null),
+                    TokenStream::from_str("super::").unwrap()
```
for consistency and simplicity, please use
`quote!{ super:: }`
rather than
`TokenStream::from_str("super::").unwrap()`
thoughout

ahl commented:
@joaommartins let me know if you'd like to get this finished up or if you'd be ok with someone else taking it over.

joaommartins commented:
> @joaommartins let me know if you'd like to get this finished up or if you'd be ok with someone else taking it over.

No need, I should have it ready soon.

ahl commented:
Excellent! Thank you for the quick reply!

ahl reviewed (approved):
made some minor updates; thanks for this

joaommartins commented:
No worries, glad to contribute.
